csharp/src/IceDiscovery/Lookup.cs

```diff
@@ -162,354 +75,240 @@ public void FindAdapterById(string domainId, string adapterId, ILookupReplyPrx?
+                    _replyAdapter.Remove(requestId);
                 }
             }
         }
     }
 
     internal class LookupReply : ILookupReply
     {
-        private readonly Lookup _lookup;
-
-        public LookupReply(Lookup lookup) => _lookup = lookup;
-
-        public void FoundObjectById(Identity id, IObjectPrx? proxy, Current c) =>
-            _lookup.FoundObject(id, c.Identity.Name, proxy!); // proxy cannot be null
-
-        public void FoundAdapterById(string adapterId, IObjectPrx? proxy, bool isReplicaGroup, Current c) =>
-            _lookup.FoundAdapter(adapterId, c.Identity.Name, proxy!, isReplicaGroup); // proxy cannot be null
-    }
-
-    internal class ObjectRequest : Request<Identity>, ITimerTask
-    {
-        public void Response(IObjectPrx proxy) => Finished(proxy);
-
-        public override void Finished(IObjectPrx? proxy)
-        {
-            foreach (TaskCompletionSource<IObjectPrx?> source in Sources)
-            {
-                source.SetResult(proxy);
-            }
-            Sources.Clear();
-        }
-
-        public void RunTimerTask() => Lookup.ObjectRequestTimedOut(this);
-
-        protected override async Task InvokeWithLookup(string domainId, ILookupPrx lookup, ILookupReplyPrx lookupReply)
-        {
-            try
-            {
-                await lookup.FindObjectByIdAsync(domainId, Id, lookupReply).ConfigureAwait(false);
-            }
-            catch (Exception ex)
-            {
-                Lookup.ObjectRequestException(this, ex);
-            }
-        }
-
-        internal ObjectRequest(Lookup lookup, Identity id, int retryCount)
-            : base(lookup, id, retryCount)
-        {
-        }
-
-    }
-
-    internal abstract class Request<T>
-    {
-        public T Id { get; }
-        public string RequestId { get; }
+        internal CancellationTokenSource CancellationSource { get; }
+        internal TaskCompletionSource<IObjectPrx?> CompletionSource { get; }
+        internal readonly HashSet<IObjectPrx> Proxies = new HashSet<IObjectPrx>();
 
-        protected List<TaskCompletionSource<IObjectPrx?>> Sources = new List<TaskCompletionSource<IObjectPrx?>>();
-        protected int FailureCount;
-        protected Lookup Lookup;
-        protected int LookupCount;
-        protected int RetryCount;
+        public void FoundObjectById(Identity id, IObjectPrx? proxy, Current current) =>
+            CompletionSource.SetResult(proxy);
 
-        public bool AddCallback(TaskCompletionSource<IObjectPrx?> source)
+        public void FoundAdapterById(string adapterId, IObjectPrx? proxy, bool isReplicaGroup, Current current)
         {
-            Sources.Add(source);
-            return Sources.Count == 1;
-        }
-
-        public bool Exception()
-        {
-            if (++FailureCount == LookupCount)
+            if (isReplicaGroup)
             {
-                Finished(null);
-                return true;
+                lock (Proxies)
```

Comment:
We should not lock an internal field like this. Lock a private mutex field (if doable), or if possible, switch to a thread-safe container.

Reply:
Ideally Proxies will be a concurrent HashSet but there isn't one in C#, I guess I can use `ConcurrentDictionary<IObjectPrx, object?>` and remove the locks, other option is to use a ConcurrentQueue/ConcurrentStack and filter the duplicates later when we create the endpoint in

https://github.com/zeroc-ice/ice/pull/882/files/34bc67daa97460a1b482c5a41e8d5533b9908cca#diff-fe34d621ab492928ed8af2d09bb8585aR186

Reply:
I updated the collection to be a private method and add an internal method for waiting for the additional replies